mm2src/coins/tendermint/tendermint_balance_events.rs

```diff
@@ -0,0 +1,161 @@
+    const EVENT_NAME: &'static str = "COIN_BALANCE";
+
+    async fn handle(self, _interval: f64) {
+        fn generate_subscription_query(query_filter: String) -> String {
+            let mut params = serde_json::Map::with_capacity(1);
+            params.insert("query".to_owned(), RpcValue::String(query_filter));
+
+            let q = MethodCall {
+                id: RpcId::Num(0),
+                jsonrpc: Some(RpcVersion::V2),
+                method: "subscribe".to_owned(),
+                params: RpcParams::Map(params),
+            };
+
+            serde_json::to_string(&q).expect("This should never happen")
+        }
+
+        let account_id = self.account_id.to_string();
+        let mut current_balances: HashMap<String, BigDecimal> = HashMap::new();
+
+        let receiver_q = generate_subscription_query(format!("coin_received.receiver = '{}'", account_id));
+        let receiver_q = tokio_tungstenite_wasm::Message::Text(receiver_q);
+
+        let spender_q = generate_subscription_query(format!("coin_spent.spender = '{}'", account_id));
+        let spender_q = tokio_tungstenite_wasm::Message::Text(spender_q);
+
+        loop {
+            let node_uri = match self.rpc_client().await {
+                Ok(client) => client.uri(),
+                Err(e) => {
+                    log::error!("{e}");
+                    continue;
+                },
+            };
+
+            let socket_address = format!("{}/{}", http_uri_to_ws_address(node_uri), "websocket");
+
+            let mut wsocket = match tokio_tungstenite_wasm::connect(socket_address).await {
+                Ok(ws) => ws,
+                Err(e) => {
+                    log::error!("{e}");
+                    continue;
+                },
+            };
+
+            // Filter received TX events
+            if let Err(e) = wsocket.send(receiver_q.clone()).await {
+                log::error!("{e}");
+                continue;
+            }
+
+            // Filter spent TX events
+            if let Err(e) = wsocket.send(spender_q.clone()).await {
+                log::error!("{e}");
+                continue;
+            }
+
+            while let Some(message) = wsocket.next().await {
+                let msg = match message {
+                    Ok(tokio_tungstenite_wasm::Message::Text(data)) => data.clone(),
+                    Ok(tokio_tungstenite_wasm::Message::Close(_)) => break,
+                    Err(err) => {
+                        log::error!("Server returned an unknown message type - {err}");
+                        break;
+                    },
+                    _ => continue,
+                };
+
+                // Here, we receive raw data from the socket.
+                // To examine this data, you can use tools like wscat/websocat or visit
+                // https://pastebin.pl/view/499cbf2c for sample data.
+                if let Ok(json_val) = serde_json::from_str::<serde_json::Value>(&msg) {
```

Comment:
Should the message string be validated/sanitized before using `serde_json::from_str`?

Reply:
How exactly? We expect JSON String value which gets validated here with `if` condition. The following logic also ensures the validity of the message.

Reply:
Yeah, this is not a problem as we don't broadcast any values we obtain from the message, I guess I was confused a bit and thought `denoms` below are used but we get the `tickers` using `denoms` and broadcast the `ticker`. https://github.com/KomodoPlatform/komodo-defi-framework/blob/0e41ae7f7599b2070bf188187a432d7dbf70c3c2/mm2src/coins/tendermint/tendermint_balance_events.rs#L92